webapp/src/app.tsx

```diff
@@ -162,10 +183,37 @@ const App = React.memo((): JSX.Element => {
                                 >
                                     <DashboardPage/>
                                 </Route>
-                                <Route path='/:boardId?/:viewId?/:cardId?'>
-                                    {loggedIn === false && <Redirect to='/login'/>}
-                                    {loggedIn === true && <BoardPage/>}
+                                <Route
+                                    exact={true}
+                                    path='/welcome'
+                                >
+                                    <WelcomePage/>
                                 </Route>
+                                <Route
+                                    path='/:boardId?/:viewId?/:cardId?'
+                                    render={({match}) => {
+                                        // Since these 3 path values are optional and they can be anything, we can pass /x/y/z and it will
+                                        // match this route however these values may not be valid so we should at the very least check
+                                        // board id for descisions made below
+                                        const boardIdIsValidUUIDV4 = (/^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$/).test(match.params.boardId || '')
+
+                                        if (loggedIn === false) {
+                                            return <Redirect to='/login'/>
+                                        }
+
+                                        if (Utils.isFocalboardPlugin() && loggedIn === true && (!localStorage.getItem('welcomePageViewed') || !boardIdIsValidUUIDV4)) {
+                                            const originalPath = `/${match.params.boardId || ''}/${match.params.viewId || ''}/${match.params.cardId || ''}`
+                                            const queryString = boardIdIsValidUUIDV4 ? `r=${originalPath}` : ''
+                                            return <Redirect to={`/welcome?${queryString}`}/>
+                                        }
+
+                                        if (loggedIn === true) {
+                                            return <BoardPage/>
+                                        }
+
+                                        return null
+                                    }}
+                                />
```

Comment:
In general, I think you should call a function for all of these cases. So that each of these routes does approximately the same thing. Something like  -

```
if (loggedIn === false) {
    return <Redirect to='/login'/>
}

if (sendtoWelcomePage()){
    redirectPage = {createRedirect} // may be a function or inline or returned from sendWelcomePage?
    return <Redirect to={`/welcome?r=${redirectPage}`}/>
}

// do whatever the default for the Route is.
//return <DashboardPage/>
return <BoardPage/>
```

Reply:
Done